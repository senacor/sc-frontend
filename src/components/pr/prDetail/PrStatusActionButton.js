import React from 'react';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles/index';

const styles = theme => ({
  buttonDesktop: {
    position: 'relative',
    backgroundColor: theme.palette.primary['400'],
    color: '#FFF',
    marginBottom: '2%'
  },
  sizeSmall: {}
});

function PrStatusActionButton(props) {
  let { classes, label } = props;
  return (
    <Button
      classes={{ sizeSmall: classes.sizeSmall }}
      size="small"
      className={classes.buttonDesktop}
      onClick={props.releaseButtonClick}
    >
      {label}
    </Button>
  );
}

export default withStyles(styles)(PrStatusActionButton);
