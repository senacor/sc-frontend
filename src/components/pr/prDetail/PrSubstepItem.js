import React from 'react';
import ListItem from '@material-ui/core/ListItem/ListItem';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import { withStyles } from '@material-ui/core';

const styles = theme => {
  return {
    listItemRoot: {
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: 0.5 * theme.spacing.unit,
      paddingBottom: 0.5 * theme.spacing.unit
    },
    listItemPrimary: {
      fontSize: '0.875rem'
    }
  };
};

const PrSubstepItem = props => {
  const { classes, substep } = props;
  return (
    <ListItem
      classes={{
        root: classes.listItemRoot
      }}
    >
      <ListItemText
        classes={{
          primary: classes.listItemPrimary
        }}
        primary={substep.label}
        secondary={
          substep.isCompleted
            ? substep.rendering.complete
            : substep.isCurrentUserActionPerformer
            ? substep.rendering.incompleteForActionPerformer
            : substep.rendering.incompleteForNonActionPerformer
        }
      />
    </ListItem>
  );
};

export default withStyles(styles)(PrSubstepItem);
