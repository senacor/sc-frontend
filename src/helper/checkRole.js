import ROLES from './roles';

export const hasRole = (role, userroles) => {
  return userroles.includes(role);
};

export const isEmployee = (userroles = []) => {
  return userroles.includes('PR_Mitarbeiter');
};

export const isSupervisor = userroles => {
  return userroles.includes(ROLES.PR_CST_LEITER);
};
