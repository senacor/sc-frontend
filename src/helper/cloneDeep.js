const cloneDeep = require('lodash/cloneDeep');

export default cloneDeep;
