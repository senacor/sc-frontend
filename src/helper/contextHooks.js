import { useContext } from 'react';
import { ErrorContext, InfoContext, UserinfoContext } from '../components/App';
import ROLES from './roles';

export const useInfoContext = () => {
  const infoContext = useContext(InfoContext.context);
  return {
    context: infoContext,
    msg: msg => {
      infoContext.setValue({
        hasInfos: true,
        messageId: msg
      });
    },
    hide: () => {
      infoContext.setValue({
        hasInfos: false,
        messageId: ''
      });
    }
  };
};

export const useUserinfoContext = () => {
  const userInfoContext = useContext(UserinfoContext.context);
  const { userroles, userinfo } = userInfoContext.value;
  return {
    context: userInfoContext,
    isSupervisorInPr: pr => {
      return userinfo.userId === pr.supervisor.id;
    },
    isReviewerInPr: pr => {
      return userinfo.userId === pr.reviewer.id;
    },
    isOwnerInPr: pr => {
      return pr.employee.id === userinfo.userId;
    },
    hasRoleAdmin: () => {
      return userroles[0] === 'ADMIN';
    },
    hasRoleHr: () => {
      return userroles.includes(ROLES.PERSONAL_DEV);
    },
    hasRoleEmployee: () => {
      return userroles.includes(ROLES.DEVELOPER);
    },
    hasRoleSupervisor: () => {
      return userroles.includes(ROLES.SUPERVISOR);
    }
  };
};

export const useErrorContext = () => {
  const errorContext = useContext(ErrorContext.context);
  return {
    context: errorContext,
    showGeneral: () => {
      errorContext.setValue({
        hasErrors: true,
        messageId: 'message.error'
      });
    },
    hide: () => {
      errorContext.setValue({ hasErrors: false, messageId: '' });
    },
    hasError: error => {
      return errorContext.value.errors && errorContext.value.errors[error];
    },
    show: (msgId, errors) => {
      errorContext.setValue(
        errors
          ? { hasErrors: true, messageId: msgId, errors: errors }
          : { hasErrors: true, messageId: msgId }
      );
    }
  };
};
