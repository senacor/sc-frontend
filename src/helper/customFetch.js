import store from '../store';
import { LOGIN_UNAUTHORIZED } from './dispatchTypes';
/**
 * supported usage:
 *   fetch('URL', configObj);
 *
 *   where configObj is something like:
 *   {
 *     method,
 *     headers,
 *     body
 *   }
 */
export default function customFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`,
    'Content-Type': 'application/json'
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}

export function fileFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}
