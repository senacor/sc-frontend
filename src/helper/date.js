import moment from 'moment';
import 'moment-timezone';

export const FRONTEND_DATE_FORMAT = 'DD.MM.YYYY';
export const FRONTEND_LOCALE_DATE_TIME_FORMAT = 'YYYY-MM-DD';

export const formatDateForFrontend = input => {
  if (input) {
    let date = moment.isMoment(input) ? input : moment(input);
    if (date.isValid()) {
      return date.format(FRONTEND_DATE_FORMAT);
    }
  }
  return null;
};

// Converting LocaleDateTime for frontend
export const formatLocaleDateTime = (input, format) => {
  if (input) {
    let modifiedInput = [...input];
    modifiedInput[1] = modifiedInput[1] - 1;
    let date = moment.isMoment(modifiedInput)
      ? modifiedInput
      : moment.utc(modifiedInput);
    if (date.isValid()) {
      return date.tz('Europe/Berlin').format(format);
    }
  }
  return null;
};

// Converting date to LocaleDateTime for backend
export const convertLocalDateTime = date => {
  const convertedDate = moment
    .tz(`${date.target.value}`, 'Europe/Berlin')
    .utc()
    .format('YYYY-MM-DDTHH:mm');
  if (convertedDate === 'Invalid date') {
    return null;
  } else return convertedDate;
};

export const isValidDate = date => {
  const isValidDate = moment(
    date,
    FRONTEND_LOCALE_DATE_TIME_FORMAT,
    true
  ).isValid();
  return isValidDate;
};

export const getReadableDate = date => {
  const minutes = date[4] > 9 ? date[4].toString() : `0${date[4]}`;
  return `${date[2]}.${date[1]}.${date[0]}, ${date[3]}:${minutes}`;
};
