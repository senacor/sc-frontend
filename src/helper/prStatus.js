export const prStatusEnum = {
  RELEASED_SHEET_REVIEWER: 'FILLED_SHEET_REVIEWER',
  RELEASED_SHEET_EMPLOYEE: 'FILLED_SHEET_EMPLOYEE',
  REQUESTED_DATE: 'REQUESTED_DATE',
  FIXED_DATE: 'ALL_DATES_ACCEPTED',
  FINALIZED_REVIEWER: 'MODIFICATIONS_ACCEPTED_REVIEWER',
  FINALIZED_EMPLOYEE: 'MODIFICATIONS_ACCEPTED_EMPLOYEE',
  ARCHIVED_HR: 'PR_COMPLETED'
};
