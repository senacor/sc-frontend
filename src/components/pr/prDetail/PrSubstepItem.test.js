import React from 'react';
import PrSubstepItem from './PrSubstepItem';
import { createShallow } from '@material-ui/core/test-utils';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import PrStatusActionButton from './PrStatusActionButton';
import { prStatusEnum } from '../../../helper/prStatus';

describe('PrSubstepItem Component', () => {
  let shallow = createShallow({ dive: true });

  const releaseButtonClickMock = jest.fn();

  const renderSubstep = {
    isCompleted: true,
    isCurrentUserActionPerformer: false,
    label: 'Mitarbeiter:',
    rendering: {
      complete: 'Abgeschlossen',
      incompleteForNonActionPerformer: 'Nicht abgeschlossen',
      incompleteForActionPerformer: (
        <PrStatusActionButton
          label={'Freigabe'}
          releaseButtonClick={releaseButtonClickMock(
            prStatusEnum.RELEASED_SHEET_EMPLOYEE
          )}
        />
      )
    }
  };

  it('should match snapshot', () => {
    const component = shallow(<PrSubstepItem substep={renderSubstep} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the correct values for a NonActionPerformer', () => {
    const component = shallow(<PrSubstepItem substep={renderSubstep} />);
    expect(
      component
        .find(ListItemText)
        .find('[primary]')
        .props().primary
    ).toEqual('Mitarbeiter:');

    expect(
      component
        .find(ListItemText)
        .find('[secondary]')
        .props().secondary
    ).toEqual('Abgeschlossen');
  });

  it('should display the correct values for an ActionPerformer', () => {
    let substep = {
      isCompleted: false,
      isCurrentUserActionPerformer: true,
      label: 'Mitarbeiter:',
      rendering: {
        complete: 'Abgeschlossen',
        incompleteForNonActionPerformer: 'Nicht abgeschlossen',
        incompleteForActionPerformer: (
          <PrStatusActionButton
            label={'Freigabe'}
            releaseButtonClick={releaseButtonClickMock(
              prStatusEnum.RELEASED_SHEET_EMPLOYEE
            )}
          />
        )
      }
    };

    const component = shallow(<PrSubstepItem substep={substep} />);
    expect(
      component
        .find(ListItemText)
        .find('[primary]')
        .props().primary
    ).toEqual('Mitarbeiter:');

    expect(
      component
        .find(ListItemText)
        .find('[secondary]')
        .props().secondary
    ).toEqual(
      <PrStatusActionButton label={'Freigabe'} releaseButtonClick={undefined} />
    );
  });
});
