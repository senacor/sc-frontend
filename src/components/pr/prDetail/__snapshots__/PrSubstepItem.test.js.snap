// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrSubstepItem Component should match snapshot 1`] = `
<WithStyles(ListItem)
  classes={
    Object {
      "root": "PrSubstepItem-listItemRoot-1",
    }
  }
>
  <WithStyles(ListItemText)
    classes={
      Object {
        "primary": "PrSubstepItem-listItemPrimary-2",
      }
    }
    primary="Mitarbeiter:"
    secondary="Abgeschlossen"
  />
</WithStyles(ListItem)>
`;
